pymeasure/adapters/adapter.py

```diff
@@ -34,44 +37,112 @@ class Adapter:
     This class should only be inherited from.
 
     :param preprocess_reply: optional callable used to preprocess strings
-        received from the instrument. The callable returns the processed string.
+        received from the instrument. The callable returns the processed
+        string. This argument is deprecated.
     :param kwargs: all other keyword arguments are ignored.
     """
 
     def __init__(self, preprocess_reply=None, **kwargs):
         self.preprocess_reply = preprocess_reply
         self.connection = None
+        self.log = logging.Logger("Adapter")
+        self.log.addHandler(logging.NullHandler())
+        if preprocess_reply is not None:
+            warn(
+                "Do not use `Adapter.values`, but `Instrument.values` instead. "
+                "Preprocess_reply is a values parameter.",
+                FutureWarning)
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
-        if self.connection is not None:
+        self.close()
+
+    def close(self):
+        """Close the connection."""
+        try:
             self.connection.close()
+        except AttributeError:
+            pass
 
-    def write(self, command):
-        """ Writes a command to the instrument
+    # Directly called methods. DO NOT OVERRIDE IN SUBCLASS!
+    def write(self, command, **kwargs):
+        """
+        Write a string command to the instrument.
 
-        :param command: SCPI command string to be sent to the instrument
+        The write_termination string is appended, if defined.
+
+        :param command: Command string to be sent to the instrument.
+        :param kwargs: Keyword arguments for the connection itself.
+        """
+        self.log.debug(("WRITE:", command))
+        self._write(command, **kwargs)
+
+    def write_bytes(self, content, **kwargs):
+        """
+        Write the bytes `content` to the instrument.
+
+        :param content: The bytes to write to the instrument.
+        :param kwargs: Keyword arguments for the connection itself.
+        """
+        self.log.debug(("WRITE:", content))
+        self._write_bytes(content, **kwargs)
+
+    def read(self, **kwargs):
+        """
+        Read until the buffer is empty and return the resulting string.
+
+        The read_termination string is removed from the response.
+
+        :param kwargs: Keyword arguments for the connection itself.
+
+        :returns: String ASCII response of the instrument.
+        """
+        read = self._read(**kwargs)
+        self.log.debug(("READ:", read))
+        return read
+
+    def read_bytes(self, count=-1, **kwargs):
+        """
+        Read a certain number of bytes from the instrument.
+
+        :param count: Number of bytes to read. For some adapters '-1' means all.
```

Comment:
"Some"?
Also, what does "all" mean -- until a timeout?

Reply:
Depends on the protocol library (pyvisa, seriel, telnet...) . Sometimes the whole buffer (you can ask for the current buffer length), or up to a timeout or up to a Termchar (option for pyvisa).
I should point explicitly to the adapter itself.

Reply:
Right now the SerialAdapter reads until a timeout. I'd change it to read_until term_char, if a term_char is defined, otherwise read until timeout.
That makes it better (you don't have to wait for the timeout, even though you already received everything you expected) and comparable to pyvisa.

With that, we can define "read" as reading until termchar (if defined) or timeout else.
A negative value ('-count') for "read_bytes" does the same with maximum 'count' bytes.

Reply:
sounds good I think